lib/Sema/SemaBounds.cpp

```diff
@@ -3285,6 +3288,318 @@ namespace {
+          if (FirstIteration) {
+            IntermediateIntersecions = GetByCFGBlock(*I)->OutElse;
+            FirstIteration = false;
+          } else
+            IntermediateIntersecions = Intersect(IntermediateIntersecions, CurrentBlock->OutElse);
+        }
+      }
+      CurrentBlock->In = IntermediateIntersecions;
+
+      // Update Out Set
+      ComparisonSet OldOutThen = CurrentBlock->OutThen, OldOutElse = CurrentBlock->OutElse;
+      CurrentBlock->OutThen = Difference(Union(CurrentBlock->In, CurrentBlock->GenThen), CurrentBlock->Kill);
+      CurrentBlock->OutElse = Difference(Union(CurrentBlock->In, CurrentBlock->GenElse), CurrentBlock->Kill);
+
+      // Recompute the Affected Blocks
+      if (Differ(OldOutThen, CurrentBlock->OutThen) || Differ(OldOutElse, CurrentBlock->OutElse))
+        for (CFGBlock::const_succ_iterator I = CurrentBlock->Block->succ_begin(), E = CurrentBlock->Block->succ_end(); I != E; ++I)
+          WorkList.push(GetByCFGBlock(*I));
+    }
+
+  //if(DEBUG_DATAFLOW)
+    for (auto B : Blocks) {
+      llvm::errs() << "Block #" << B->Block->getBlockID() << ":\n";
+      B->Block->dump();
+      llvm::errs() << "{\n";
+      PrintComparisonSet(B->In, std::string("In"));
+      PrintComparisonSet(B->OutThen, std::string("OutThen"));
+      PrintComparisonSet(B->OutElse, std::string("OutElse"));
+      llvm::errs() << "}\n";
+    }
+  }
+
+  void Reset() {
+    CurrentIndex = 0;
+  }
+
+  void Next() {
+    CurrentIndex++;
+  }
+
+  // This function fills `ComparisonFacts` with pairs (Expr1, Expr2) where
+  // Expr1 < Expr2, Expr1 <= Expr2, Expr2 > Expr1, or Expr2 >= Expr1.
+  // These comparisons correspond to the current block.
+  void GetFacts(std::set<std::pair<Expr *, Expr *>>& ComparisonFacts) {
+    ComparisonFacts = Blocks[CurrentIndex]->In;
+  }
+
+private:
+  ElevatedCFGBlock* GetByCFGBlock(const CFGBlock *B) {
+    return Blocks[BlockIDs[B->getBlockID()]];
+  }
+
+  // Given two sets S1 and S2, the return value is S1 \ S2.
+  ComparisonSet Difference(ComparisonSet& S1, ComparisonSet& S2) {
+   if (S2.size() == 0)
+      return S1;
+    ComparisonSet Result;
+    for (auto E1 : S1)
+      if (S2.find(E1) == S2.end())
+        Result.insert(E1);
+    return Result;
+  }
+
+  // Given two sets S1 and S2, the return value is the union of these sets.
+  ComparisonSet Union(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() == 0)
+      return S2;
+    if (S2.size() == 0)
+      return S1;
+    ComparisonSet Result(S1);
+    Result.insert(S2.begin(), S2.end());
+    return Result;
+  }
+
+  // Given two sets S1 and S2, this function returns true if the two sets
+  // are not equal.
+  // Equality means that their sizes are the same, and every member of S1 is found in
+  // S2 and vice versa.
+  bool Differ(ComparisonSet& S1, ComparisonSet& S2) {
+    if (S1.size() != S2.size())
+      return true;
+    if (S1.size() == 0 && S2.size() == 0)
+      return false;
+    if (S1.size() > S2.size())
```

Comment:
This logic seems wrong.  By the first check in the function, S1.size() must equal S2.size().